runtime/runc_jailer.go

```diff
@@ -0,0 +1,351 @@
+
+			// link drives to new contents path
+			for i, d := range m.Cfg.Drives {
+				drivePath := firecracker.StringValue(d.PathOnHost)
+				fileName := filepath.Base(drivePath)
+				newDrivePath := filepath.Join(rootPath, fileName)
+
+				if err := os.Link(drivePath, newDrivePath); err != nil {
+					return errors.Wrapf(err, "failed to copy drive %v", drivePath)
+				}
+
+				f, err := os.Open(drivePath)
+				if err != nil {
+					return errors.Wrap(err, "failed to open drive file")
+				}
+
+				if internal.IsStubDrive(f) {
+					j.logger.Debug("stub drive detected")
+					if err := os.Chown(newDrivePath, int(j.uid), int(j.gid)); err != nil {
+						if e := f.Close(); e != nil {
+							j.logger.WithError(e).Debug("failed to close drive file")
+						}
+						return errors.Wrap(err, "failed to change ownership of stub drive")
+					}
+				}
+
+				if err := f.Close(); err != nil {
+					j.logger.WithError(err).Debug("failed to close drive file")
+				}
+
+				j.logger.WithField("drive", newDrivePath).Debug("Adding drive")
+				m.Cfg.Drives[i].PathOnHost = firecracker.String(fileName)
+			}
+
+			// Setting the proper path to where the vsock path should be
+			for i, v := range m.Cfg.VsockDevices {
+				filename := filepath.Base(v.Path)
+				v.Path = filepath.Join("/", filename)
+				m.Cfg.VsockDevices[i] = v
+			}
+
+			j.logger.Info("Successfully ran jailer handler")
+			return nil
+		},
+	}
+}
+
+// BuildLinkFifoHandler will return a new firecracker.Handler with the function
+// that will allow linking of the fifos making them visible to Firecracker.
+func (j runcJailer) BuildLinkFifoHandler() firecracker.Handler {
+	return firecracker.Handler{
+		Name: jailerFifoHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			contentsPath := j.RootPath()
+			fifoFileName := filepath.Base(m.Cfg.LogFifo)
+			newFifoPath := filepath.Join(contentsPath, fifoFileName)
+			if err := os.Link(m.Cfg.LogFifo, newFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.LogFifo = newFifoPath
+
+			metricFifoFileName := filepath.Base(m.Cfg.MetricsFifo)
+			newMetricFifoPath := filepath.Join(contentsPath, metricFifoFileName)
+			if err := os.Link(m.Cfg.MetricsFifo, newMetricFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.MetricsFifo = newMetricFifoPath
+
+			return nil
+		},
+	}
+}
+
+// ExposeDeviceToJail will inspect the given file, srcFilePath, and based on the
+// file type, proper handling will occur to ensure that the file is visible in
+// the jail. For block devices we will use mknod to create the device and then
+// set the correct permissions to ensure visibility in the jail.
+func (j runcJailer) ExposeDeviceToJail(srcMountPath string) error {
+	uid := j.uid
+	gid := j.gid
+
+	stat := syscall.Stat_t{}
+	if err := syscall.Stat(srcMountPath, &stat); err != nil {
+		return err
+	}
+
+	// Checks file type using S_IFMT which is the bit mask for the file
+	// type. Here we only care about block devices, ie S_IFBLK.
```

Comment:
Maybe the method should be renamed to `ExposeBlockDeviceToJail`?

Reply:
My thinking was if we ever cared for other file types and not just `S_IFBLK`, it wouldn't be too much of a hassle of just adding the additional type, and the method wouldn't need to be renamed.